projects/ADuCM3029_demo_cn0414/src/adc_update_timer.c

```diff
@@ -0,0 +1,271 @@
+ * value of 32.
+ *
+ * @param [in/out] dev            - Update timer descriptor structure. Contains
+ *        the desired frequency as input parameter and the software prescaler as
+ *        output parameter.
+ * @param [out] ptr_tmr_load_val  - New load value of the timer.
+ * @param [out] ptr_tmr_prescaler - New timer prescaler value.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int32_t adc_update_timer_get_config(struct adc_update_desc *dev,
+		uint16_t *ptr_tmr_load_val, ADI_TMR_PRESCALER *ptr_tmr_prescaler)
+{
+	uint32_t pclk_val;
+	float sw_divided_freq, prescaled_freq;
+	uint8_t i, j;
+	uint8_t done_flag;
+	uint32_t ret;
+
+	for(j = 1; j < 33; j++) {
+		dev->sw_prescaler = j;
+
+		ret = adi_pwr_GetClockFrequency(ADI_CLOCK_PCLK, &pclk_val);
+		if(ret != ADI_PWR_SUCCESS)
+			return ret;
+
+		sw_divided_freq = pclk_val / dev->sw_prescaler;
+		for(i = 0; i < 4; i++) {
+			if(i == 0)
+				prescaled_freq = sw_divided_freq / (dev->f_update * 8);
+			else
+				prescaled_freq = sw_divided_freq /
+						 (pow(4, i + 1) * (dev->f_update * 8));
+			if(prescaled_freq < 0xFFFF) {
+				*ptr_tmr_load_val = (uint16_t)prescaled_freq;
+				*ptr_tmr_prescaler = i;
+				done_flag = 1;
+				break;
+			}
+		}
+		if(done_flag == 1)
+			break;
+	}
+
+	return ret;
+}
+
+/**
+ * Initializes the timer which updates the ADC input registers.
+ *
+ * @param [out] device    - The device structure.
+ * @param [in] init_param - Pointer to initialization structure.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t adc_update_setup(struct adc_update_desc **device,
+			 struct adc_update_init *init_param)
+{
+	int32_t ret;
+	struct adc_update_desc *dev;
+	ADI_TMR_CONFIG timer_config;
+
+	dev = malloc(sizeof(*dev));
+	if(!dev)
+		ret = -1;
+
+	dev->f_update = init_param->f_update;
+	dev->update_timer = init_param->update_timer;
+
+	/* Basic timer configuration */
+	timer_config.bCountingUp = false;
+	timer_config.bPeriodic = true;
+	timer_config.bReloading = false;
+	timer_config.bSyncBypass = true;
+	timer_config.eClockSource = ADI_TMR_CLOCK_PCLK;
+
+	ret = adc_update_timer_get_config(dev, &timer_config.nLoad,
+					  &timer_config.ePrescaler);
+	if(ret != 0)
+		goto error;
+	timer_config.nAsyncLoad = timer_config.nLoad;
+
+	ret = adi_tmr_Init(dev->update_timer, timer0_adc_update_callback,
+			   NULL, true);
+	if(ret != ADI_TMR_SUCCESS)
+		goto error;
+
+	ret = adi_tmr_ConfigTimer(dev->update_timer, &timer_config);
+	if(ret != ADI_TMR_SUCCESS)
+		goto error;
+
+	*device = dev;
+
+	return ret;
+
+error:
+	free(dev);
+
+	return ret;
+}
+
+/**
+ * Stop timer and free the resources allocated by adc_update_remove().
+ *
+ * @param [in] dev - The device structure.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t adc_update_remove(struct adc_update_desc *dev)
+{
+	int32_t ret;
+
+	if(!dev)
+		return -1;
+
+	ret = adc_update_activate(dev, false);
+	if(ret != 0)
```

Comment:
nitpick: `if(ret)` or `if(ret != SUCCESS)`, since you are using this everywhere in your code